### PR TITLE
v0.0.1 Release

### PR DESCRIPTION
**Pull request by CurtisLusmore**
Added the following key functions:
* New-AwsAccount
* Edit-AwsAccount
* New-AwsRole
* Set-AwsProfile
* New-AwsSession

### CODE REVIEW

File: PwshAwsConfig.psm1 [hunk 1]
```diff
@@ -0,0 +1,420 @@
+<#
+.SYNOPSIS
+A collection of functions for managing AWS CLI config and credentials.
+
+.DESCRIPTION
+A collection of functions for managing AWS CLI config and credentials.
```
Comment (scottjbaldwin) [inline PwshAwsConfig.psm1:6]: Might want to dive a bit deeper in the description.

File: PwshAwsConfig.psm1 [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+Create IAM credentials for a new AWS account.
+
+.DESCRIPTION
+Create IAM credentials for a new AWS account. This will create a record in your
+~/.aws/credentials file to represent an IAM user in the given AWS account, by
+reading the AWS Access Key ID and AWS Secret Access Key from a console prompt
+or CSV file. This will silently ignore if the account already exists. To edit
+an existing account, use Edit-AwsAccount.
+
+.PARAMETER AccountName
```
Comment (scottjbaldwin) [inline PwshAwsConfig.psm1:20]: The naming might need some work. From an AWS perspective the concept of an `Account` is very well defined, but this does not map one-to-one to an "AWS Account". Not sure what other naming you could use, maybe `Organisation`, or `Realm` or `Identity`???? I guess this comment also applies to the Function name too. It may just be a case of it' too hard to name it anything other than Account.

**Comment by CurtisLusmore** [comment]
I should update the README as well

File: PwshAwsConfig.psm1 [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+        [Parameter(Mandatory)] [string] $AccountName,
+        [Parameter(Mandatory)] [string] $RoleName,
+        [Parameter(Mandatory)] [string] $RoleArn,
+        [switch] $UseIam
+    )
+
+    If ($UseIam)
+    {
+        $sourceProfile = "$AccountName`:iam"
+        If (-not (Test-AwsProfile -Profile $sourceProfile))
```
Comment (scottjbaldwin) [inline PwshAwsConfig.psm1:168]: Can this not be checked outside the `If ($UseIam)` by setting the `$sourceProfile` in an `If`?

File: PwshAwsConfig.psm1 [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+        $sourceProfile = "$AccountName`:mfa"
+        If (-not (Test-AwsProfile -Profile $sourceProfile))
+        {
+            throw "No existing account $AccountName exists. Use New-AwsAccount to create a new account"
+        }
+    }
+
+    $profile = "$AccountName`:$RoleName"
+
+    aws configure --profile $profile set role_arn $RoleArn
```
Comment (scottjbaldwin) [inline PwshAwsConfig.psm1:184]: Given all of the operations ultimately rely on the aws cli being installed, it might be useful to do a pre-emptive check on this dependency.

File: PwshAwsConfig.psm1 [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+#>
+Function Set-AwsProfile
+{
+    Param(
+        [string] $Profile
+    )
+
+    If (-not $Profile)
+    {
+        $hotkeys = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
```
Comment (scottjbaldwin) [inline PwshAwsConfig.psm1:224]: I know it's a horrible thought to entertain, but what happens if there are more than 72 profiles? Possibly a V2.0 thing consideration :).

File: PwshAwsConfig.psm1 [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+The example below creates a new record in your ~/.aws/credentials file for
+work:iam by reading the AWS Access Key ID and AWS Secret Access Key from the
+supplied CSV file.
+
+PS C:> New-AwsAccount -AccountName work -AccessKeyCsvPath my.user_accessKeys.csv
+#>
+Function New-AwsAccount
+{
+    Param(
+        [Parameter(Mandatory)] [string] $AccountName,
```
Comment (scottjbaldwin) [inline PwshAwsConfig.psm1:46]: One thing I like to do in scripts that will be used a lot is to make them positional parameters where possible (e.g. `[Parameter(Mandatory=$true, Position=0)]` etc...), that way a power user could just go `New-AwsAccount vicroads accesskeys.csv` rather than having to type in out the parameter names in full to disambiguate.

File: PwshAwsConfig.psm1 [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+
+PS C:> New-AwsSession
+MFA Code: ******
+#>
+Function New-AwsSession
+{
+    Param(
+        [string] $Profile = $env:AWS_PROFILE,
+        [string] $Code,
+        [string] $Arn
```
Comment (scottjbaldwin) [inline PwshAwsConfig.psm1:284]: Same comment about positional parameters as above

**Review by scottjbaldwin — APPROVED** [review]

File: PwshAwsConfig.psd1 [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+FunctionsToExport = @(
+    'New-AwsAccount',
+    'Edit-AwsAccount',
+    'New-AwsRole',
+    'Set-AwsProfile',
+    'New-AwsSession'
+)
+
+# Cmdlets to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no cmdlets to export.
+CmdletsToExport = @()
```
[the remark this answers is not in the document]
Comment (CurtisLusmore) [reply PwshAwsConfig.psd1:83]: I... don't know. Honestly I barely know the difference between a function and a cmdlet.

File: PwshAwsConfig.psd1 [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+# Private data to pass to the module specified in RootModule/ModuleToProcess. This may also contain a PSData hashtable with additional module metadata used by PowerShell.
+PrivateData = @{
+
+    PSData = @{
+
+        # Tags applied to this module. These help with module discovery in online galleries.
+        # Tags = @()
+
+        # A URL to the license for this module.
+        # LicenseUri = ''
```
[the remark this answers is not in the document]
Comment (CurtisLusmore) [reply PwshAwsConfig.psd1:109]: Good question, I'll ping some peeps and see what we should use. MIT would be nice.

File: PwshAwsConfig.psm1 (hunk 1 above)
Comment (CurtisLusmore) [reply PwshAwsConfig.psm1:6]: Yep good idea. I'll be filling out the README too so likely to be a lot of overlap in content here.

File: PwshAwsConfig.psm1 (hunk 4 above)
Comment (CurtisLusmore) [reply PwshAwsConfig.psm1:184]: Great point, I might address this separately because there's a bunch of ways we could approach this.

File: PwshAwsConfig.psm1 (hunk 5 above)
Comment (CurtisLusmore) [reply PwshAwsConfig.psm1:224]: 62* but yep, good point. I think at that point the interactive menu becomes overwhelming anyway and maybe it needs some kind of search feature 😅

File: PwshAwsConfig.psm1 (hunk 6 above)
Comment (CurtisLusmore) [reply PwshAwsConfig.psm1:46]: Happy to add this explicitly, although I think this Just Works™ naturally by defaulting to the order listed.

File: PwshAwsConfig.psm1 (hunk 3 above)
Comment (CurtisLusmore) [reply PwshAwsConfig.psm1:168]: Yes I think you're right, these two branches are identical.

File: PwshAwsConfig.psd1 (hunk 8 above)
[the remark this answers is not in the document]
Comment (scottjbaldwin) [reply PwshAwsConfig.psd1:83]: My understanding is that a CmdLet is a Function with special powers... when you add the attribute `[CmdletBinding()]` This turns an ordinary function into a commandlet that can support things like `-Verbose` `-WhatIf` etc... without the attribute, it's just a plain old Function.

[the remark this answers is not in the document]
Comment (RobFaie) [reply PwshAwsConfig.psd1:83]: in the module manifest the `CmdletsToExport` refers to binary CMDlets that have been compiled from C# and available in a linked dll.

The `FunctionsToExport` is for script functions regardless of whether they do or do not have the `[CMDletBinding()]`or not

File: PwshAwsConfig.psd1 (hunk 9 above)
[the remark this answers is not in the document]
Comment (RobFaie) [reply PwshAwsConfig.psd1:109]: We use MIT for our existing powershell modules and is the preferred license in general

File: PwshAwsConfig.psm1 [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+    If (Test-AwsProfile -Profile $credentialName)
+    {
+        Write-Information "Account already exists"
+        Return
+    }
+
+    NewOrEdit-AwsAccount -AccountName $AccountName -AccessKeyCsvPath $AccessKeyCsvPath
+}
+
+<#
```
Comment (RobFaie) [inline PwshAwsConfig.psm1:61]: I think comment based help needs to be within the function block before the `Param(`bit in order to surface properly. You can check by loading the module and calling `Get-Help Edit-AwsAccount`

**Review by RobFaie — APPROVED** [review]

File: PwshAwsConfig.psm1 (hunk 2 above)
Comment (CurtisLusmore) [reply PwshAwsConfig.psm1:20]: Yeah I reckon `Identity` or `User` would do the trick.

File: PwshAwsConfig.psd1 [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+# Type files (.ps1xml) to be loaded when importing this module
+# TypesToProcess = @()
+
+# Format files (.ps1xml) to be loaded when importing this module
+# FormatsToProcess = @()
+
+# Modules to import as nested modules of the module specified in RootModule/ModuleToProcess
+NestedModules = @(
+    'PwshAwsConfig.psm1'
```
[the remark this answers is not in the document]
Comment (CurtisLusmore) [reply PwshAwsConfig.psd1:70]: Cool thanks, fixed!

File: PwshAwsConfig.psm1 (hunk 10 above)
Comment (CurtisLusmore) [reply PwshAwsConfig.psm1:61]: Wow interesting, I had only ever seen it like this, before the `Function` keyword. It turns out it can go in 3 places!

https://docs.microsoft.com/en-us/powershell/scripting/developer/help/placing-comment-based-help-in-functions?view=powershell-7.1

File: PwshAwsConfig.psm1 [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+    Param(
+        [Parameter(Mandatory)] [string] $AccountName,
+        [string] $AccessKeyCsvPath
+    )
+
+    $credentialName = "$AccountName`:iam"
+
+    If (Test-AwsProfile -Profile $credentialName)
+    {
+        Write-Information "Account already exists"
```
Comment (vatsalyagoel) [inline PwshAwsConfig.psm1:54]: Maybe reword to Profile Already Exists
